core/common/src/main/java/alluxio/underfs/Fingerprint.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.Constants;
+
+import com.google.common.base.Splitter;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Fingerprint for a UFS file or directory.
+ */
+@NotThreadSafe
+public abstract class Fingerprint {
```

Comment:
Why do we need separate `DirectoryFingerprint` and `FileFIngerprint` classes? It adds complexity to need to coordinate between the superclass and subclass. Could we have a single `Fingerprint` class with an optional `contentHash` field? We can introduce an additional tag to track whether the inode is a file or a directory.